### PR TITLE
refactor: update BFF view logic

### PR DESCRIPTION
**Pull request by nsprenkle**
**TL;DR -** changed how we call `get_learner_data` into 2 specific modes, with and without a step suffix.

JIRA: [AU-1544](https://2u-internal.atlassian.net/browse/AU-1544)

**What changed?**

- When calling `get_learner_data` without a step, get updated progress but no responses. Used by XBlock view and in between steps in the popout to refresh progress.
- When calling `get_learner_data` with a step, validates that you _can_ get to that specific step (have reached / surpassed that step) and, if so, passes the appropriate response data.

**Other changes**

- Renamed some functions and context objects.
- More consistent error presentation from errors in `get_learner_data`.
- Organized and extended error codes.
- Updated some docstrings.

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:

- [ ] I've done a visual code review
- [ ] I've tested the new functionality

FYI: @openedx/content-aurora

### CODE REVIEW

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 1]
```diff
@@ -62,61 +63,49 @@ def get_block_info(self, data, suffix=""):  # pylint: disable=unused-argument
...
-        # Allow jumping to a specific step, within our allowed steps
-        if jump_step:
-            jumpable_steps = ("submission", "peer", "grades")
-            if jump_step not in jumpable_steps:
-                raise JsonHandlerError(404, f"Invalid jump to step: {jump_step}")
-            if self._can_jump_to_step(workflow_step, jump_step):
-                serializer_context.update({"jump_to_step": suffix})
-            else:
-                raise JsonHandlerError(400, f"Cannot jump to step: {jump_step}")
+        active_step = suffix
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/mixin.py:73]: I think this should be moved down to line 85 for clarity

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 2]
```diff
@@ -62,61 +63,49 @@ def get_block_info(self, data, suffix=""):  # pylint: disable=unused-argument
...
-            "submission": "submission",
-            "peer": "peer",
-            "grades": "done"
-        }
+        # Check to see if user can access this step
+        if not self._user_has_reached_step(workflow_step, active_step):
+            raise OraApiException(400, error_codes.INACCESSIBLE_STEP, f"Inaccessible step: {active_step}")
+
+        # If they have access to this step, return the associated data
+        serializer_context.update({"active_step": active_step})
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/mixin.py:90]: nit: this can just be serializer_context["active_step"] = active_step, we're just doing the one field

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 3]
```diff
@@ -62,61 +63,49 @@ def get_block_info(self, data, suffix=""):  # pylint: disable=unused-argument
...
-            - Check if we can get to that step
-            - If we can, go to that step
-            - Otherwise, raise an error
+        Get data for the user / step of the ORA, based on the following modes:
+
+        1) If no step provided, refresh progress but don't return any response.
+        2) If step provided, validate that we can get to that step and return the appropriate response
         """
+        # Query workflow step here only once to avoid duplicate workflow updates
         workflow_step = self.workflow_data.status or "submission"
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/mixin.py:72]: nit: variable names? 

workflow_step / active_step are very similar and easy to get mixed up which is which. maybe current_workflow_step and requested_step?

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 4]
```diff
@@ -62,61 +63,49 @@ def get_block_info(self, data, suffix=""):  # pylint: disable=unused-argument
...
-        }
+        # Check to see if user can access this step
+        if not self._user_has_reached_step(workflow_step, active_step):
+            raise OraApiException(400, error_codes.INACCESSIBLE_STEP, f"Inaccessible step: {active_step}")
+
+        # If they have access to this step, return the associated data
+        serializer_context.update({"active_step": active_step})
+        return PageDataSerializer(self, context=serializer_context).data
+
+    def _user_has_reached_step(self, workflow_step, step):
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/mixin.py:93]: This is a similar nit as above, but these variable names are all very similar and I'd suggest maybe some clarification

File: openassessment/xblock/ui_mixins/mfe/mixin.py [hunk 5]
```diff
@@ -62,61 +63,49 @@ def get_block_info(self, data, suffix=""):  # pylint: disable=unused-argument
...
-        # Can always "jump" to the step you're on
-        if jump_step == workflow_step:
+        # Have reached your current workflow step
+        if requested_workflow_step == workflow_step:
             return True
 
-        # Can jump to a step you've completed
-        step_status = self.workflow_data.status_details.get(workflow_step_to_jump_to, {})
+        # Have reached any step you have completed
+        step_status = self.workflow_data.status_details.get(requested_workflow_step, {})
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/mixin.py:108]: workflow api method that takes a step name and returns whether or not it has been completed?

File: openassessment/xblock/ui_mixins/mfe/page_context_serializer.py [hunk 6]
```diff
@@ -245,7 +246,7 @@ class ProgressSerializer(Serializer):
 
     def get_activeStepName(self, instance):
         """Return the active step name"""
-        if not instance.workflow_data.has_workflow:
+        if not instance.workflow_data.has_workflow or instance.workflow_data.is_cancelled:
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/page_context_serializer.py:249]: Could you explain this one? Do we not have a UI view for "cancelled"?

File: openassessment/xblock/ui_mixins/mfe/page_context_serializer.py [hunk 7]
```diff
@@ -298,40 +314,32 @@ def get_response(self, instance):
...
+            learner_submission_data = instance.get_learner_submission_data()
+            return SubmissionSerializer(learner_submission_data).data
 
-        else:
-            raise Exception("Missing view context for page")
+        # Steps without a necessary response
+        elif active_step in ("staff"):
+            response = None
+
+        return AssessmentResponseSerializer(response).data
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/page_context_serializer.py:342]: Nice.Good job cleaning this up

**Review by jansenk — APPROVED** [review]
Looks great, just some small stylistic nits and suggestions for clearing up confusions I had while reading through it

**Review by leangseu-edx — APPROVED** [review]
LGTM
